### PR TITLE
Register training sheet for equipment

### PR DESCRIPTION
**Pull request by Lan2u**
Functionally:
Add the ability to register a training form/id for a piece of equipment.

Actually:
Try and workout what is going on. 

My biggest question is around 'events'. Is the idea that everything is represented as a history/list of events and everything is worked out from that?

Would you expect to be able to do?:
```typescript
function unwrap<T>(val: O.Option<T>): T {
    if (O.isSome(val)) {
        return val.value;
    }
    throw new Error('Failed to unwrap');
}

function callCommand<T> (func: Command<T>["process"], command: T) {
    return (events: ReadonlyArray<DomainEvent>) => RA.append(
        unwrap(func({
            command,
            events,
        }))
    )(events);
}

function createEquipment(events: ReadonlyArray<DomainEvent>){
    const areaName = faker.commerce.productName() as NonEmptyString;
    const areaDescription = faker.commerce.productDescription();
    const areaId = v4() as UUID;
    const equipmentId = v4() as UUID;
    return {
        events: pipe(
            events,
            callCommand(create.process, {
                id: areaId,
                name: areaName,
                description: areaDescription,
            }),
            RA.map(e => { console.log(e); return e; }),
            callCommand(add.process, {
                id: equipmentId,
                name: faker.commerce.productName() as NonEmptyString,
                areaId: areaId,
            }),
            RA.map(e => { console.log(e); return e; }),
        ),
        equipmentId,
    };
}
```

I tried this and it didn't work because add.process uses RA.match which seems to only construct an event if the existing events are empty?

Maybe I misunderstand the point of events / how its supposed to be used?

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -4,7 +4,6 @@ check: test lint typecheck unused-exports
 
 node_modules: package.json bun.lockb
 	bun install --frozen-lockfile
-	touch node_modules
```
Comment (Lan2u) [inline Makefile:7]: Right now I've dev'ing on windows so dropped this

File: src/read-models/equipment/get-all.ts [hunk 2]
```diff
@@ -23,6 +24,12 @@ export const getAll = (
         O.map(area => ({
           ...equipment,
           areaName: area.name,
```
Comment (erkannt) [inline src/read-models/equipment/get-all.ts:26]: At some point we probably should refactor this to stop repeatedly going over all events. As events grow this kind of shape will kill our performance.
For now lets now worry about.

Comment (Lan2u) [reply src/read-models/equipment/get-all.ts:26]: The current setup/architecture seems to revolve around repeatedly going through events - the fact this isn't particularly performant was my first ? when I looked through it.

I assume the idea behind the architecture is that:
- the simplicity in storing the data
- looking at the history
- adding new events
- benefits of using a read-only functional approach
- very easy testing

outweights the
1. additional complexity for simple operations like reading/updating the current state
2. performance impacts of repeatedly copying / making new arrays
3. performance impacts of looping through all events

with the potential mitigations of
1. using lots of helpers from RA / ones we write ourselves
3. index the events

2. is just a drawback of trying to do readonly stuff in javascript/typescript which, unlike a proper compiled language like rust, cannot make high performance optimisations

**Comment by Lan2u** [comment]
> The idea is to follow CQRS principles and separate the concerns of
> commands from queries.
> In this case that means:
> - it is the read-models responsibility to deal with e.g. training sheet
>   registrations for equipment that does not exist
> - 'resources' are only allowed to care about events that they raise
> - the 'process' function of a resource may never fail it can either
>   produce an event or not
> - 'process' responsibilities are mostly to ensure idempotency
> - turning the history of events into a meaningful and consistent
>   respresentation for users is a query/read-model responsibility
> 
> The 'unwrap' like functionality is implemented as 'getSomeOrFail' and
> 'getRightOrFail' for use in tests.

**Pull request merged**